Pytorch module

kohr-h opened the pull request:
This adds missing functionality for proper pytorch integration. The existing `autograd.Function` wrapper cannot be used directly since `pytorch` doesn't quite know how to deal with extra axes when the Function doesn't handle that, and a naive implementation results in gradients not being propagated beyond that layer.
This PR fixes that and adds examples and tests for the new feature.

Some things are also renamed and moved around.

adler-j commented on odl/contrib/pytorch/operator.py:
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        op_name = self.op_func.operator.__class__.__name__
```
Why not repr the actual operator?

kohr-h replied:
Because of the self-printing of composed modules. For example:
```
NNFBP (
  (conv_det): Conv2d(1, 32, kernel_size=(1, 11), stride=(1, 1), padding=(0, 10))
  (backproj): OperatorAsModule(RayBackProjection) ((50, 256) -> (128, 128))
  (conv1_vol): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(2, 2))
  (conv2_vol): Conv2d(32, 1, kernel_size=(3, 3), stride=(1, 1), padding=(2, 2))
)
```
But I can check if that works with `__str__`.

adler-j commented on odl/contrib/pytorch/test/test_operator.py:
```diff
@@ -0,0 +1,191 @@
+shape = simple_fixture('shape', [(3,), (2, 3), (2, 2, 3)])
+
+
+def test_autograd_function_forward(dtype, use_cuda):
```
Very nice tests!

adler-j commented on odl/contrib/pytorch/test/test_operator.py:
```diff
@@ -0,0 +1,191 @@
+    res_var = torch_op(x_var)
+
+    # ODL result
+    odl_res = odl_op(x.cpu().numpy())
```
We should verify that if `use_cuda` is true, the result is still on CUDA

adler-j commented:
By the way, have you managed to re-implement the learned primal dual article using pytorch?

kohr-h commented on odl/contrib/pytorch/operator.py:
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
+        x : `torch.autograd.variable.Variable`
+            Input of this layer. The contained tensor must have shape
+            ``extra_shape + operator.domain.shape``, and
+            ``len(extra_shape)`` must be at least 1 (batch axis).
```
That would just be an extra axis. This module is a bit like a linear layer that only understands input shape, output shape and batch axis. But in contrast to the linear layer, which can literally only act with one extra axis, this one supports any number of extra axes (batch, channel, you name it), simply by flattening them and then looping. Since there are no parameters to learn, any extra axis looks the same.

adler-j replied:
This made 100% sense, try to get some of that into the doc!

kohr-h commented on odl/contrib/pytorch/examples/operator_autograd.py:
```diff
@@ -1,9 +1,15 @@
 import torch
 import odl
-from odl.contrib.pytorch import TorchOperator
+from odl.contrib import pytorch
```
Okay, I'll go for something like `odl_torch` then. Had myself a bit doubts about `pytorch` :-)

kohr-h commented on odl/contrib/pytorch/examples/operator_module.py:
```diff
@@ -0,0 +1,85 @@
+"""Demonstration of ODL operators as pytorch modules.
+
+This example shows how to wrap an ODL ``Operator`` as a
+``torch.nn.Module`` that can be used as a layer in a Neural Network.
```
You mean the contrast to `autograd.Function`?

kohr-h commented on odl/contrib/pytorch/operator.py:
```diff
@@ -63,40 +73,52 @@ def forward(self, input):
             self.save_for_backward(input)
 
-        input_arr = input.numpy()
+        # TODO: use GPU memory directly if possible
```
Yeah well, for that the latter needs to exist :-)

kohr-h commented on odl/contrib/pytorch/operator.py:
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
+        ...                    [0, 1, 1]], dtype='float32')
+        >>> odl_op = odl.MatrixOperator(matrix)
+        >>> torch_op = OperatorAsModule(odl_op)
+        >>> torch_op.op_func.operator is odl_op
```
OK

kohr-h commented on odl/contrib/pytorch/operator.py:
```diff
@@ -222,9 +245,162 @@ def backward(self, grad_output):
+        (1 ,.,.) =
+           2   4
+           3   6
+        <BLANKLINE>
```
Dunno, I do want to have a non-square batch shape, and I think 3x2x2 is the smallest I can get. The `<BLANKLINE>` things are not displayed in the rendered docs btw. They're just for doctest.

adler-j commented:
Really minor stuff left to get this in, lets get it done.

kohr-h commented:
Yes, I'll look into this tomorrow when I'm back in office. I don't feel like compiling pytorch again on my laptop :-)

kohr-h commented on odl/contrib/torch/examples/operator_module.py:
```diff
@@ -0,0 +1,89 @@
+print('Operator layer evaluated on a 1x3 tensor variable:')
+print(op_layer(inp))
+
+inp = autograd.Variable(torch.ones(3))[None, None, ...]
```
I don't know if examples are the place to do this. Extra batch and channel axes is all that people likely need to use, so I think I'll keep it at this.

kohr-h commented:
Rebased and fixed the stuff. There was something I wanted to fix as well but don't remember now. But I'm gonna merge this now and fix the other things in the next PR.
